### PR TITLE
BigInt: implement add, subtract

### PR DESCRIPTION
**Pull request by duongnhn**
- implement add, subtract
- throw type error if type mismatch
- assign with add, subtract

### CODE REVIEW

File: test/BigInt/subtract.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+//-------------------------------------------------------------------------------------------------------
+
+
+if (this.WScript && this.WScript.LoadScriptFile) { // Check for running in ch
+    this.WScript.LoadScriptFile("..\\UnitTestFramework\\UnitTestFramework.js");
+}
+
+var tests = [
+    {
+        name: "Add BigInt literal",
```
Comment (sethbrenith) [inline test/BigInt/subtract.js]: Nit: a few of these test names could change from Add to Subtract

File: lib/Runtime/Math/JavascriptMath.cpp [hunk 2]
```diff
@@ -227,6 +227,15 @@ using namespace Js;
             Js::TypeId typeLeft = JavascriptOperators::GetTypeId(aLeft);
             Js::TypeId typeRight = JavascriptOperators::GetTypeId(aRight);
 
+            if (typeLeft == TypeIds_BigInt || typeRight == TypeIds_BigInt)
```
Comment (sethbrenith) [inline lib/Runtime/Math/JavascriptMath.cpp]: Could this go in JavascriptMath::Add_FullHelper right before the call to Add_Helper? That way you don't have to duplicate it because both of these methods end up calling that helper function. Also, I expect the profiling compiler will already do this for us, but ideally we would check for more common cases (numbers and strings) first.

File: lib/Runtime/Library/JavascriptBigInt.cpp [hunk 3]
```diff
@@ -397,4 +414,183 @@ namespace Js
         return (leftBigInt->Compare(rightBigInt) == 0);
     }
 
+    // pbi1 += pbi2 assume pbi1 > pbi2
```
Comment (sethbrenith) [inline lib/Runtime/Library/JavascriptBigInt.cpp]: If there's an assumption on the input data, it's nice to add it as an Assert within the top of the method.

File: lib/Runtime/Library/JavascriptBigInt.cpp [hunk 4]
```diff
@@ -397,4 +414,183 @@ namespace Js
...
+    {
+        digit_t carryDigit = 0;
+        digit_t *pDigit1 = pbi1->m_digits;
+        digit_t *pDigit2 = pbi2->m_digits;
+        digit_t i = 0;
+
+        for (; i < pbi2->m_length; i++)
+        {
+            digit_t tempCarryDigit = 0;
+            if (pDigit2[i] > 0)
```
Comment (sethbrenith) [inline lib/Runtime/Library/JavascriptBigInt.cpp]: Do you know if these `> 0` checks save a lot of time? If not, or if you're unsure, I'd prefer to omit them and make the code more compact.

File: lib/Runtime/Library/JavascriptBigInt.cpp [hunk 5]
```diff
@@ -397,4 +414,183 @@ namespace Js
...
+                pDigit1[i] = JavascriptBigInt::AddDigit(pDigit1[i], carryDigit, &tempCarryDigit);
+            }
+            carryDigit = tempCarryDigit;
+        }
+
+        if (0 < carryDigit) // length increase by 1
+        {
+            if (pbi1->m_length >= pbi1->m_maxLength && !pbi1->Resize(pbi1->m_length + 1))
+            {
+                return;
```
Comment (sethbrenith) [inline lib/Runtime/Library/JavascriptBigInt.cpp]: I see a couple of different behaviors in this file for how we handle failure to resize. I don't see this case included in the spec (it usually ignores anything to do with the reality of running on a physical machine with less than infinite memory). I could imagine three possible options:

* Return a wrong answer (what we do here)
* Throw an exception
* Crash the program (what we do on line 217)

I'd like to propose that we update Resize to return void and throw a `RangeError` (`JavascriptError::ThrowRangeError`) if it can't resize, since this seems most consistent with other parts of the spec, such as the behaviors when a program tries to create a string or array that is longer than allowed.

File: lib/Runtime/Library/JavascriptBigInt.cpp [hunk 6]
```diff
@@ -397,4 +414,183 @@ namespace Js
...
+
+        for (; i < pbi1->m_length; i++)
+        {
+            digit_t tempCarryDigit = 0;
+            if (carryDigit > 0)
+            {
+                pDigit1[i] = JavascriptBigInt::SubDigit(pDigit1[i], carryDigit, &tempCarryDigit);
+            }
+            carryDigit = tempCarryDigit;
+            if (carryDigit == 0)
```
Comment (sethbrenith) [inline lib/Runtime/Library/JavascriptBigInt.cpp]: This could go in the test clause of the `for` loop, to keep the control flow simpler. And AddAbsolute could use a similar check.

File: lib/Runtime/Library/JavascriptBigInt.cpp [hunk 7]
```diff
@@ -397,4 +414,183 @@ namespace Js
...
+            pbi1->m_length--;
+        }
+        Assert(pbi1->m_length > 0);
+    }
+
+    JavascriptBigInt * JavascriptBigInt::Sub(JavascriptBigInt * pbi1, JavascriptBigInt * pbi2)
+    {
+        if (JavascriptBigInt::IsZero(pbi1))
+        {
+            pbi2->m_isNegative = !pbi2->m_isNegative;
```
Comment (sethbrenith) [inline lib/Runtime/Library/JavascriptBigInt.cpp]: pbi2 shouldn't be modifiable at this point. We need to clone it before negating.

Comment (sethbrenith) [reply lib/Runtime/Library/JavascriptBigInt.cpp]: The `const` keyword would probably be useful in preventing this sort of bug. These top-level methods could take `const JavascriptBigInt*`, and then to get a copy that the compiler will let you modify, you'd have to call `JavascriptBigInt::New`.

---
In reply to: [227860626](https://github.com/Microsoft/ChakraCore/pull/5792#discussion_r227860626) [](ancestors = 227860626)

Comment (sethbrenith) [reply lib/Runtime/Library/JavascriptBigInt.cpp]: Also probably worth including some cases in assign_by_value.js that verify that the second arg to addition and subtraction doesn't get modified.

---
In reply to: [227861565](https://github.com/Microsoft/ChakraCore/pull/5792#discussion_r227861565) [](ancestors = 227861565,227860626)

File: lib/Runtime/Library/JavascriptBigInt.cpp [hunk 8]
```diff
@@ -397,4 +414,183 @@ namespace Js
...
+            }
+        }
+        return pbi1;
+    }
+
+    Var JavascriptBigInt::Add(Var aLeft, Var aRight)
+    {
+        JavascriptBigInt *leftBigInt = VarTo<JavascriptBigInt>(aLeft);
+        JavascriptBigInt *rightBigInt = VarTo<JavascriptBigInt>(aRight);
+        return JavascriptBigInt::Add(JavascriptBigInt::New(leftBigInt, leftBigInt->GetScriptContext()), JavascriptBigInt::New(rightBigInt, rightBigInt->GetScriptContext()));
```
Comment (sethbrenith) [inline lib/Runtime/Library/JavascriptBigInt.cpp]: Currently this does two allocations, and ideally we would need only one to create the output BigInt (or zero in the special case where one of the args is zero, because we can return the other arg directly by reference). I don't think we're worried about fine-tuning perf at this point, so this is fine, but maybe it would be worth adding a comment with something like "Consider deferring creation of new instances until we need them".

File: lib/Runtime/Library/JavascriptBigInt.cpp (hunk 3 above)
Comment (duongnhn) [reply lib/Runtime/Library/JavascriptBigInt.cpp]: should we add it if that checking is expensive?

Comment (sethbrenith) [reply lib/Runtime/Library/JavascriptBigInt.cpp]: Yeah, `Assert` is only included in debug builds so it's fine to do expensive stuff there. In test or release builds, it goes away.

---
In reply to: [227876426](https://github.com/Microsoft/ChakraCore/pull/5792#discussion_r227876426) [](ancestors = 227876426)

File: lib/Runtime/Math/JavascriptMath.cpp (hunk 2 above)
Comment (duongnhn) [reply lib/Runtime/Math/JavascriptMath.cpp]: @sethbrenith I need to duplicate it here because we have 2 different functions for x64 and x86. I think I can put it in inline JavascriptMath::Add but it is "inline" so I am not sure if I should.

File: test/BigInt/subtract.js (hunk 1 above)
Comment (duongnhn) [reply test/BigInt/subtract.js]: will do

File: lib/Runtime/Math/JavascriptMath.cpp (hunk 2 above)
Comment (sethbrenith) [reply lib/Runtime/Math/JavascriptMath.cpp]: There are two different methods here based on `#if FLOATVAR`, but they both end up calling Add_FullHelper_Wrapper, which calls Add_FullHelper. So I think Add_FullHelper is a reasonable place to add this so we only have to do it once.

Also, I just realized there's also a correctness concern. We're supposed to call ToPrimitive on the inputs, which Add_FullHelper_Wrapper does. Consider: `1n + {valueOf: () => 2n}`. According to the spec we're supposed to coerce the second arg to a primitive, which calls its `valueOf` method. I think your current implementation would throw. This is also a problem for subtraction.

So basically, I think the right way to think about this method is that it's optimized fast paths for some common cases, but the more complete code belongs in Add_FullHelper.

---
In reply to: [227879760](https://github.com/Microsoft/ChakraCore/pull/5792#discussion_r227879760) [](ancestors = 227879760)

File: lib/Runtime/Library/JavascriptBigInt.cpp (hunk 8 above)
Comment (duongnhn) [reply lib/Runtime/Library/JavascriptBigInt.cpp]: will add TODO

File: lib/Runtime/Library/JavascriptBigInt.cpp (hunk 5 above)
Comment (duongnhn) [reply lib/Runtime/Library/JavascriptBigInt.cpp]: will do

File: lib/Runtime/Library/JavascriptBigInt.cpp [hunk 9]
```diff
@@ -397,4 +414,183 @@ namespace Js
...
+            JavascriptBigInt::AddAbsolute(pbi1, pbi2);// -a-b=-(a+b)
+            return pbi1;
+        }
+        else // both positive 
+        {
+            if (pbi1->CompareAbsolute(pbi2) == 0) // a -a = 0
+            {
+                return JavascriptBigInt::CreateZero(pbi1->GetScriptContext());
+            }
+            else if (pbi1->CompareAbsolute(pbi2) == 1)
```
[the remark this answers is not in the document]
Comment (duongnhn) [reply lib/Runtime/Library/JavascriptBigInt.cpp]: will use switch

File: lib/Runtime/Library/JavascriptBigInt.cpp (hunk 4 above)
Comment (duongnhn) [reply lib/Runtime/Library/JavascriptBigInt.cpp]: will remove

File: lib/Runtime/Library/JavascriptBigInt.cpp (hunk 6 above)
Comment (duongnhn) [reply lib/Runtime/Library/JavascriptBigInt.cpp]: will do

**Comment by duongnhn** [comment]
@dotnet-bot test OSX _no_jit_shared_osx_osx_test please

File: lib/Runtime/Math/JavascriptMath.cpp (hunk 2 above)
Comment (sethbrenith) [reply lib/Runtime/Math/JavascriptMath.cpp]: You don't need to implement valueOf; that's part of ToPrimitive. The change you made should in ToPrmitive is enough; I just verified on your latest iteration that `1n + {valueOf: () => 2n} == 3n`.

File: lib/Runtime/Library/JavascriptBigInt.cpp [hunk 10]
```diff
@@ -397,4 +412,170 @@ namespace Js
...
+            digit_t tempCarryDigit = 0;
+            pDigit1[i] = JavascriptBigInt::SubDigit(pDigit1[i], pDigit2[i], &tempCarryDigit);
+            pDigit1[i] = JavascriptBigInt::SubDigit(pDigit1[i], carryDigit, &tempCarryDigit);
+            carryDigit = tempCarryDigit;
+        }
+
+        for (; i < pbi1->m_length && carryDigit > 0; i++)
+        {
+            digit_t tempCarryDigit = 0;
+            if (carryDigit > 0)
```
Comment (sethbrenith) [inline lib/Runtime/Library/JavascriptBigInt.cpp]: One last `if` we could remove

File: lib/Runtime/Math/JavascriptMath.cpp [hunk 11]
```diff
@@ -807,6 +818,18 @@ using namespace Js;
         Var JavascriptMath::Subtract_Full(Var aLeft, Var aRight, ScriptContext* scriptContext)
         {
             JIT_HELPER_REENTRANT_HEADER(Op_Subtract_Full);
+            Js::TypeId typeLeft = JavascriptOperators::GetTypeId(aLeft);
```
Comment (sethbrenith) [inline lib/Runtime/Math/JavascriptMath.cpp]: We're supposed to call ToNumeric on each input before checking their types.

Comment (duongnhn) [reply lib/Runtime/Math/JavascriptMath.cpp]: but ToNumber return a double?

File: lib/Runtime/Math/JavascriptMath.cpp (hunk 2 above)
Comment (duongnhn) [reply lib/Runtime/Math/JavascriptMath.cpp]: I'm happy to add this to test :)

File: lib/Runtime/Math/JavascriptMath.cpp (hunk 11 above)
Comment (sethbrenith) [reply lib/Runtime/Math/JavascriptMath.cpp]: Not ToNumber, ToNumeric is a new method defined by the BigInt spec.

Comment (duongnhn) [reply lib/Runtime/Math/JavascriptMath.cpp]: are you suggesting rewriting the whole subtract flow? or just call toPrimitive?

Technically, if we want to follow strictly to spec then we need to rewrite the whole addition flow too.

Comment (sethbrenith) [reply lib/Runtime/Math/JavascriptMath.cpp]: Calling ToPrimitive is plenty, this looks good to me.

---
In reply to: [228003513](https://github.com/Microsoft/ChakraCore/pull/5792#discussion_r228003513) [](ancestors = 228003513)

File: test/BigInt/add.js [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+        body: function () {
+            var x = 3n;
+            var y = 2n;
+            y += x;
+            assert.isTrue(x == 3n);
+            assert.isTrue(y == 5n);
+            y = x + 4n;
+            assert.isTrue(y == 7n);
+            y = 8n + x;
+            assert.isTrue(y == 11n);
```
Comment (boingoing) [inline test/BigInt/add.js]: Are BigInt primitives supported for strict equality ?

Comment (sethbrenith) [reply test/BigInt/add.js]: not yet
